README.md
## linux_networking (noetic) - 1.1.0-1

The packages in the `linux_networking` repository were released into the `noetic` distro by running `/usr/bin/bloom-release linux_networking --rosdistro noetic --track noetic` on `Tue, 02 Mar 2021 01:32:38 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.1.0-1`

Versions of tools used:

- bloom version: `0.10.0`
- catkin_pkg version: `0.4.23`
- rosdep version: `0.20.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## linux_networking (melodic) - 1.0.16-1

The packages in the `linux_networking` repository were released into the `melodic` distro by running `/usr/bin/bloom-release linux_networking --rosdistro melodic --track melodic` on `Fri, 08 Nov 2019 12:41:21 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: https://github.com/pr2-gbp/linux_networking-release.git
- rosdistro version: `1.0.13-2`
- old version: `1.0.15-0`
- new version: `1.0.16-1`

Versions of tools used:

- bloom version: `0.8.0`
- catkin_pkg version: `0.4.12`
- rosdep version: `0.15.2`
- rosdistro version: `0.7.4`
- vcstools version: `0.1.42`


## linux_networking (lunar) - 1.0.15-0

The packages in the `linux_networking` repository were released into the `lunar` distro by running `/usr/bin/bloom-release --rosdistro lunar --track lunar linux_networking --edit` on `Sat, 30 Mar 2019 22:50:40 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: https://github.com/pr2-gbp/linux_networking-release.git
- rosdistro version: `1.0.12-0`
- old version: `1.0.12-0`
- new version: `1.0.15-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.11`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.3`
- vcstools version: `0.1.40`


## linux_networking (melodic) - 1.0.15-0

The packages in the `linux_networking` repository were released into the `melodic` distro by running `/usr/bin/bloom-release --track melodic --ros-distro melodic linux_networking --edit` on `Sat, 30 Mar 2019 17:51:51 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: https://github.com/pr2-gbp/linux_networking-release.git
- rosdistro version: `1.0.13-2`
- old version: `1.0.13-2`
- new version: `1.0.15-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.11`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.3`
- vcstools version: `0.1.40`


## linux_networking (kinetic) - 1.0.15-0

The packages in the `linux_networking` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --track kinetic --ros-distro kinetic linux_networking` on `Tue, 26 Mar 2019 07:30:53 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: https://github.com/pr2-gbp/linux_networking-release.git
- rosdistro version: `1.0.14-0`
- old version: `1.0.14-0`
- new version: `1.0.15-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (kinetic) - 1.0.14-0

The packages in the `linux_networking` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --track kinetic --ros-distro kinetic linux_networking --edit` on `Mon, 18 Mar 2019 19:02:00 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: https://github.com/pr2-gbp/linux_networking-release.git
- rosdistro version: `1.0.12-0`
- old version: `1.0.12-0`
- new version: `1.0.14-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (melodic) - 1.0.13-2

The packages in the `linux_networking` repository were released into the `melodic` distro by running `/usr/bin/bloom-release --rosdistro melodic --track melodic linux_networking` on `Thu, 07 Mar 2019 07:32:05 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.13-1`
- new version: `1.0.13-2`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.0`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (melodic) - 1.0.13-1

The packages in the `linux_networking` repository were released into the `melodic` distro by running `/usr/bin/bloom-release --rosdistro melodic --track melodic linux_networking` on `Thu, 07 Mar 2019 07:23:18 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.13-0`
- new version: `1.0.13-1`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (melodic) - 1.0.13-0

The packages in the `linux_networking` repository were released into the `melodic` distro by running `/usr/bin/bloom-release --rosdistro melodic --track melodic linux_networking` on `Thu, 07 Mar 2019 07:08:04 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.13-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (lunar) - 1.0.12-0

The packages in the `linux_networking` repository were released into the `lunar` distro by running `/usr/bin/bloom-release --rosdistro lunar --track lunar linux_networking` on `Sat, 02 Mar 2019 05:48:42 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.12-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.0`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (kinetic) - 1.0.12-0

The packages in the `linux_networking` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --rosdistro kinetic --track kinetic linux_networking --edit` on `Tue, 26 Feb 2019 21:49:40 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:

- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.12-0`

Versions of tools used:

- bloom version: `0.7.2`
- catkin_pkg version: `0.4.10`
- rosdep version: `0.15.1`
- rosdistro version: `0.7.2`
- vcstools version: `0.1.40`


## linux_networking (indigo) - 1.0.10-0

The packages in the `linux_networking` repository were released into the `indigo` distro by running `/usr/bin/bloom-release --rosdistro indigo --track indigo linux_networking` on `Wed, 16 Sep 2015 15:19:41 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:
- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.10-0`

Versions of tools used:
- bloom version: `0.5.20`
- catkin_pkg version: `0.2.8`
- rosdep version: `0.11.2`
- rosdistro version: `0.4.2`
- vcstools version: `0.1.36`


## linux_networking (hydro) - 1.0.5-0

The packages in the `linux_networking` repository were released into the `hydro` distro by running `/usr/bin/bloom-release --rosdistro hydro --track hydro linux_networking` on `Mon, 06 Oct 2014 15:56:01 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:
- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: https://github.com/TheDash/linux_networking-release.git
- rosdistro version: `1.0.4-0`
- old version: `1.0.4-0`
- new version: `1.0.5-0`

Versions of tools used:
- bloom version: `0.5.12`
- catkin_pkg version: `0.2.4`
- rosdep version: `0.10.30`
- rosdistro version: `0.3.5`
- vcstools version: `0.1.35`


## linux_networking (hydro) - 1.0.4-0

The packages in the `linux_networking` repository were released into the `hydro` distro by running `/usr/bin/bloom-release --rosdistro hydro --track hydro linux_networking --edit` on `Fri, 03 Oct 2014 15:40:53 -0000`

These packages were released:
- `access_point_control`
- `asmach`
- `asmach_tutorials`
- `ddwrt_access_point`
- `hostapd_access_point`
- `ieee80211_channels`
- `linksys_access_point`
- `linux_networking`
- `multi_interface_roam`
- `network_control_tests`
- `network_detector`
- `network_monitor_udp`
- `network_traffic_control`

Version of package(s) in repository `linux_networking`:
- upstream repository: https://github.com/PR2/linux_networking.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.4-0`

Versions of tools used:
- bloom version: `0.5.12`
- catkin_pkg version: `0.2.4`
- rosdep version: `0.10.30`
- rosdistro version: `0.3.5`
- vcstools version: `0.1.35`


linux_networking-release
========================

Release repository for https://github.com/pr2/linux_networking


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package network_monitor_udp
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.1.0 (2021-03-02)
------------------
* Merge pull request `#6 <https://github.com/pr2/linux_networking/issues/6>`_ from k-okada/add_noetic_travis
* add more from __future_\_ import print_function
* fix indent/space mixture
* use 2to3 to convert print/except/raise/lambda for python3
* Contributors: Kei Okada

1.0.16 (2019-11-08)
-------------------
* ROS-melodic .debs for network_monitor_udp incomplete (`#3 <https://github.com/pr2/linux_networking/issues/3>`_)
  * update melodic-devel to include appropriate files
  - python modules were missing (needed setup.py and `catkin_python_setup`) from .deb
  - python executables were missing (needed `catkin_install_python`) from .deb
  - python package `roslib` is deprecated; `rostime` now lives in `rospy`
  - compiled binaries and .launch file were missing from from .deb
* Contributors: David Feil-Seifer, bk-mtg

1.0.15 (2019-03-19)
-------------------

1.0.13 (2019-03-18)
-------------------
* fixed unsigned char errors for melodic compile
* Contributors: David Feil-Seifer
* updated cmake for message build dependency issue in kinetic
* Contributors: David Feil-Seifer

1.0.12 (2019-02-26)
-------------------
* changes to cmakelists files to make packages compile in kinetic
* Contributors: David Feil-Seifer

1.0.9 (2014-10-14)
------------------

1.0.8 (2014-10-10)
------------------
* Removed rosbuild files
* Contributors: TheDash

1.0.7 (2014-10-10)
------------------

1.0.6 (2014-10-10)
------------------

1.0.5 (2014-10-06)
------------------

1.0.4 (2014-10-03)
------------------

1.0.3 (2014-10-03)
------------------

1.0.2 (2014-10-03)
------------------


nodes/roam_node.py
#! /usr/bin/env python

from __future__ import with_statement
from __future__ import print_function

import roslib; roslib.load_manifest('multi_interface_roam') 
import rospy
import std_msgs.msg
import multi_interface_roam.multi_interface_roam as mir
import multi_interface_roam.rosmasterless
import sys
import os
import os.path
import time
import yaml
from diagnostic_msgs.msg import DiagnosticStatus, KeyValue, DiagnosticArray
from pr2_msgs.msg import AccessPoint

freq_to_chan_map = {
    2412000000 : 1, 
    2417000000 : 2, 
    2422000000 : 3, 
    2427000000 : 4, 
    2432000000 : 5, 
    2437000000 : 6, 
    2442000000 : 7, 
    2447000000 : 8, 
    2452000000 : 9, 
    2457000000 : 10, 
    2462000000 : 11, 
    2467000000 : 12, 
    2472000000 : 13, 
    2484000000 : 14, 
    3657500000 : 131, 
    3662500000 : 132, 
    3660000000 : 132, 
    3667500000 : 133, 
    3665000000 : 133, 
    3672500000 : 134, 
    3670000000 : 134, 
    3677500000 : 135, 
    3682500000 : 136, 
    3680000000 : 136, 
    3687500000 : 137, 
    3685000000 : 137, 
    3689500000 : 138, 
    3690000000 : 138, 
    4915000000 : 183, 
    4920000000 : 184, 
    4925000000 : 185, 
    4935000000 : 187, 
    4940000000 : 188, 
    4945000000 : 189, 
    4960000000 : 192, 
    4980000000 : 196, 
    5035000000 : 7, 
    5040000000 : 8, 
    5045000000 : 9, 
    5055000000 : 11, 
    5060000000 : 12, 
    5080000000 : 16, 
    5170000000 : 34, 
    5180000000 : 36, 
    5190000000 : 38, 
    5200000000 : 40, 
    5210000000 : 42, 
    5220000000 : 44, 
    5230000000 : 46, 
    5240000000 : 48, 
    5260000000 : 52, 
    5280000000 : 56, 
    5300000000 : 60, 
    5320000000 : 64, 
    5500000000 : 100, 
    5520000000 : 104, 
    5540000000 : 108, 
    5560000000 : 112, 
    5580000000 : 116, 
    5600000000 : 120, 
    5620000000 : 124, 
    5640000000 : 128, 
    5660000000 : 132, 
    5680000000 : 136, 
    5700000000 : 140, 
    5745000000 : 149, 
    5765000000 : 153, 
    5785000000 : 157, 
    5805000000 : 161, 
    5825000000 : 165, 
}

class MultiInterfaceDiagPublisher:
    def __init__(self, config_file):
        with open(config_file, 'r') as f:
           self.config = yaml.load(f.read())
        if 'wireless_namespace' in self.config:
            self.wireless_namespace = self.config['wireless_namespace']
        else:
            self.wireless_namespace = 'wifi'

        try:
            mir.logdir = self.config['log_directory']
        except KeyError:
            pass

#        console_output_file = os.path.join(mir.logdir, "console_output.log") 
#        try:
#            if self.config['redirect_console']:
#                print "Redirecting output to:", console_output_file
#                sys.stdout = open(console_output_file, "a", 1)
#                sys.stderr = sys.stdout
#        except KeyError:
#            pass
#        except IOError:
#            print "Error redirecting output to: ", console_output_file

        self.diag_pub = rospy.Publisher("/diagnostics", DiagnosticArray)
        self.wifi_pub = rospy.Publisher(self.wireless_namespace+"/accesspoint", AccessPoint)
        wireless_interfaces = []
        interfaces = self.config['interfaces'] 
        for i in interfaces:
            if interfaces[i]['type'] == 'wireless':
                wireless_interfaces.append(i)
        self.wifi_sub_pub = dict((iface, rospy.Publisher(self.wireless_namespace+"/"+iface+"/accesspoint", AccessPoint)) for iface in wireless_interfaces)

        self.hostname = os.uname()[1]
        print()
        print("**************************************************************")
        print(mir.log_time_string(time.time()), "restarting.")
        print("**************************************************************")
        mir.main(config_file, mir.SimpleSelectionStrategy, self.publish_diags)
        print("**************************************************************")
        print("Exiting.")
        print("**************************************************************")
        print()
        rospy.signal_shutdown("main has exited")

    @staticmethod
    def fill_diags(name, summary, hid, diags):
        ds = DiagnosticStatus()
        ds.values = [KeyValue(k, str(v)) for (k, v) in diags]
        ds.hardware_id = hid
        ds.name = rospy.get_caller_id().lstrip('/') + ": " + name
        ds.message = summary
        return ds
    
    @staticmethod
    def frequency_to_channel(freq):
        # A bit horrible, but will do until the message changes.
        try:
            return freq_to_chan_map[freq]
        except:
            return -1

    @staticmethod
    def gen_accesspoint_msg(iface):
        msg = AccessPoint()
        msg.essid = iface.essid
        msg.macaddr = iface.bssid
        msg.signal = iface.wifi_signal 
        msg.noise = iface.wifi_noise
        msg.snr = msg.signal - msg.noise
        msg.quality = iface.wifi_quality
        msg.rate = iface.wifi_rate
        msg.tx_power = iface.wifi_txpower
        msg.channel = MultiInterfaceDiagPublisher.frequency_to_channel(iface.wifi_frequency)
        return msg
    
    def publish_diags(self, strategy):
        now = rospy.get_rostime()
        ns = strategy.ns
        ds = self.fill_diags("synthetic interface", ns.diag_summary, self.hostname, strategy.ns.diags)
        ds.level = ns.diag_level
        statuses = [ds]

        for i in range(0, len(ns.interfaces)):
            iface = ns.interfaces[i]
            ds = self.fill_diags(iface.iface, iface.diag_summary, self.hostname, iface.diags)
            statuses.append(ds)
            if iface.__class__ == mir.WirelessInterface:
                msg = self.gen_accesspoint_msg(iface)
                msg.header.stamp = now
                self.wifi_sub_pub[iface.iface].publish(msg)
                if i == ns.active_iface:
                    self.wifi_pub.publish(msg)

        da = DiagnosticArray()
        da.header.stamp = rospy.get_rostime()
        da.status = statuses
        self.diag_pub.publish(da)
        
if __name__ == "__main__":
    if len(sys.argv) != 2:
      print(file=sys.stderr)
      print("usage: roam_node.py config_file.yaml", file=sys.stderr)
      print(file=sys.stderr)
      print("This node cannot presently be used with roslaunch as it will be confused by the __name and __log parameters.", file=sys.stderr)
      sys.exit(1)
    
    rospy.init_node("multi_interface_roam", disable_rosout=True, disable_rostime=True, disable_signals=True)

    MultiInterfaceDiagPublisher(sys.argv[1])
